### PR TITLE
tests: bsim_bt: Extend connection test case.

### PR DESCRIPTION
**Pull request by wopu-ot**
Make peripheral role part of test case to build it with
split/non-split stack. Change connection parameters and channel map
after connection is established.

Signed-off-by: Wolfgang Puffitsch <wopu@oticon.com>

### CODE REVIEW

File: tests/bluetooth/bsim_bt/bsim_test_app/src/test_connect1.c [hunk 1]
```diff
@@ -268,8 +275,7 @@ static void device_found(const bt_addr_le_t *addr, s8_t rssi, u8_t type,
 
 	/* We're only interested in connectable events */
 	if (type == BT_LE_ADV_IND || type == BT_LE_ADV_DIRECT_IND) {
-		/* TODO: Move this to a place it can be shared */
-		ad_parse(ad, eir_found, (void *)addr);
+		bt_data_parse(ad, eir_found, (void *)addr);
```
Comment (jhedberg) [inline tests/bluetooth/bsim_bt/bsim_test_app/src/test_connect1.c]: This refactoring to use bt_data_parse() seems unrelated to the rest of the patch. Could you move that into a separate commit please? (can still be part of this same PR).

**Review by jhedberg — CHANGES_REQUESTED** [review]

File: tests/bluetooth/bsim_bt/bsim_test_app/src/test_connect1.c (hunk 1 above)
Comment (wopu-ot) [reply tests/bluetooth/bsim_bt/bsim_test_app/src/test_connect1.c]: Yes, I will do that.

File: tests/bluetooth/bsim_bt/bsim_test_app/src/test_connect2.c [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+		return;
+	}
+
+	bt_conn_cb_register(&conn_callbacks);
+	bt_conn_auth_cb_register(&auth_cb_display);
+
+	/* Implement notification. At the moment there is no suitable way
+	 * of starting delayed work so we do it here
+	 */
+	for (int i = 0; i < WAIT_TIME-1; i++) {
```
Comment (aescolar) [inline tests/bluetooth/bsim_bt/bsim_test_app/src/test_connect2.c]: It feels a bit too unreliable to have the test wait for WAIT_TIME-1 iterations before passing, as WAIT_TIME was just a timeout to fail the test.
Would it make sense to have this loop be an infinite loop, counting notifications sent,
and have the test pass after having sent a predefined number of notifications?

File: tests/bluetooth/bsim_bt/bsim_test_app/tests_scripts/basic_conn.sh [hunk 3]
```diff
@@ -24,12 +24,13 @@ BOARD="${BOARD:-nrf52_bsim}"
 
 cd ${BSIM_OUT_PATH}/bin
 
-Execute ./bs_${BOARD}_samples_bluetooth_peripheral_prj_conf \
-  -v=${VERBOSITY_LEVEL} -s=${SIMULATION_ID} -d=0 -RealEncryption=0
+Execute ./bs_${BOARD}_tests_bluetooth_bsim_bt_bsim_test_app_prj_conf \
+  -v=${VERBOSITY_LEVEL} -s=${SIMULATION_ID} -d=0 -RealEncryption=0 \
+  -testid=peripheral -rs=6
```
Comment (aescolar) [inline tests/bluetooth/bsim_bt/bsim_test_app/tests_scripts/basic_conn.sh]: Seeding both devices with the same random seed seems a bit funny (same in the other testcase script below)

File: tests/bluetooth/bsim_bt/bsim_test_app/src/test_connect2.c [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+		/* Battery level simulation */
+		bas_notify();
+	}
+
+	/* We consider it passed */
+	bst_result = Passed;
+	printk("Testcase passed\n");
+
+	/* Wait to be killed */
+	for (;;) {
```
Comment (aescolar) [inline tests/bluetooth/bsim_bt/bsim_test_app/src/test_connect2.c]: Doing it this way (in both sides) the devices will never end the simulation, but you rely on the sim_length passed to the Phy.
(This Phy timeout is 20 simulated seconds now)

If, in this device, all you want is to see it send 1 notification to pass, then you could have it change the test status to passed after the first notification, and just let this device go on sending notifications every second.
You could then let the other side device be the one that ends the simulation (with the same condition it had before).

**Review by aescolar — CHANGES_REQUESTED** [review]

File: tests/bluetooth/bsim_bt/bsim_test_app/tests_scripts/basic_conn_ecrypted.sh [hunk 5]
```diff
@@ -24,12 +24,13 @@ BOARD="${BOARD:-nrf52_bsim}"
 
 cd ${BSIM_OUT_PATH}/bin
 
-Execute ./bs_${BOARD}_samples_bluetooth_peripheral_prj_conf \
-  -v=${VERBOSITY_LEVEL} -s=${SIMULATION_ID} -d=0 -RealEncryption=1
+Execute ./bs_${BOARD}_tests_bluetooth_bsim_bt_bsim_test_app_prj_conf \
+  -v=${VERBOSITY_LEVEL} -s=${SIMULATION_ID} -d=0 -RealEncryption=0 \
```
[the remark this answers is not in the document]
Comment (aescolar) [reply tests/bluetooth/bsim_bt/bsim_test_app/tests_scripts/basic_conn_ecrypted.sh]: Ok.. it is not passing :)
```
d_01: @00:00:02.366132  [SUBSCRIBED]
d_00: @00:00:02.465774  [00:00:02.465,759] <err> bt_smp: reason 0x8
d_01: @00:00:03.041006  [NOTIFICATION] data 0x5674df4b length 2
d_01: @00:00:04.047255  [NOTIFICATION] data 0x5674df4b length 2
d_01: @00:00:04.047255  Testcase passed
d_00: @00:00:05.039520  Testcase passed
d_00: @00:00:09.122272  Disconnected (reason 8)
```

[the remark this answers is not in the document]
Comment (wopu-ot) [reply tests/bluetooth/bsim_bt/bsim_test_app/tests_scripts/basic_conn_ecrypted.sh]: I reworked the logic to use callbacks, so encryption should work now.

File: tests/bluetooth/bsim_bt/bsim_test_app/src/test_connect2.c (hunk 2 above)
Comment (wopu-ot) [reply tests/bluetooth/bsim_bt/bsim_test_app/src/test_connect2.c]: It makes sense, and I updated accordingly.

File: tests/bluetooth/bsim_bt/bsim_test_app/tests_scripts/basic_conn.sh (hunk 3 above)
Comment (wopu-ot) [reply tests/bluetooth/bsim_bt/bsim_test_app/tests_scripts/basic_conn.sh]: No need to do funny stuff, I use a different seed now for the peripheral.

File: tests/bluetooth/bsim_bt/bsim_test_app/src/test_connect2.c [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+static void test_con2_main(void)
+{
+	static int notify_count;
+	int err;
+
+	err = bt_enable(bt_ready);
+	if (err) {
+		printk("Bluetooth init failed (err %d)\n", err);
```
Comment (aescolar) [inline tests/bluetooth/bsim_bt/bsim_test_app/src/test_connect2.c]: Should all these error conditions we treated as test case failures? (bst_result = Failed, and bs_trace_error_line() )

File: tests/bluetooth/bsim_bt/bsim_test_app/src/test_connect1.c [hunk 7]
```diff
@@ -83,8 +97,18 @@ static u8_t notify_func(struct bt_conn *conn,
...
+		/* Disconnect before actually passing */
+		err = bt_conn_disconnect(default_conn, BT_HCI_ERR_SUCCESS);
+		if (err) {
+			printk("Disconnection failed (err %d)\n", err);
+			return BT_GATT_ITER_STOP;
+		}
+
 		bst_result = Passed;
-		bs_trace_exit_time("Testcase passed\n");
+		printk("Testcase passed\n");
```
Comment (aescolar) [inline tests/bluetooth/bsim_bt/bsim_test_app/src/test_connect1.c]: Silly question: Why this change?

File: tests/bluetooth/bsim_bt/bsim_test_app/tests_scripts/basic_conn.sh [hunk 8]
```diff
@@ -24,12 +24,13 @@ BOARD="${BOARD:-nrf52_bsim}"
 
 cd ${BSIM_OUT_PATH}/bin
 
-Execute ./bs_${BOARD}_samples_bluetooth_peripheral_prj_conf \
-  -v=${VERBOSITY_LEVEL} -s=${SIMULATION_ID} -d=0 -RealEncryption=0
+Execute ./bs_${BOARD}_tests_bluetooth_bsim_bt_bsim_test_app_prj_conf \
+  -v=${VERBOSITY_LEVEL} -s=${SIMULATION_ID} -d=0 -RealEncryption=0 \
+  -testid=peripheral -rs=23
```
Comment (aescolar) [inline tests/bluetooth/bsim_bt/bsim_test_app/tests_scripts/basic_conn.sh]: Just a comment: You do not necessarily need to set the random seed to anything if you don't want. The random seed should be default be set to a value calculated from the device number (-d)

**Review by aescolar — CHANGES_REQUESTED** [review]
It is way nicer now. The reject is only because I think failing the test on all those error checks would be worthwhile

File: tests/bluetooth/bsim_bt/bsim_test_app/src/test_connect2.c (hunk 6 above)
Comment (wopu-ot) [reply tests/bluetooth/bsim_bt/bsim_test_app/src/test_connect2.c]: At least some of the error conditions made the test case fail implicitly by not proceeding. However, I now updated the pass/fail criteria to explicitly fail in case of error conditions and make this clear.

**Review by aescolar — APPROVED** [review]

File: tests/bluetooth/bsim_bt/bsim_test_app/src/test_connect1.c (hunk 1 above)
Comment (wopu-ot) [reply tests/bluetooth/bsim_bt/bsim_test_app/src/test_connect1.c]: @jhedberg The commit has been split accordingly. Do you have any further comments?

File: tests/bluetooth/bsim_bt/bsim_test_app/src/test_connect2.c [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+ * Basic connection test:
+ *   We expect a central to connect to us.
+ *
+ *   The thread code is mostly a copy of the peripheral_hr sample device
+ */
+
+#define WAIT_TIME 5 /*seconds*/
+extern enum bst_result_t bst_result;
+
+#define FAIL(...)					\
```
Comment (jhedberg) [inline tests/bluetooth/bsim_bt/bsim_test_app/src/test_connect2.c]: How come you're redefining these here? Wouldn't it make sense to have them in some local header file?

**Review by jhedberg — APPROVED** [review]
Looks ok to me, except the the seemingly unnecessary redefinition of PASS/FAIL (which I don't consider a blocker, so approving)
